packages/vite/src/node/plugins/asset.ts

```diff
@@ -344,18 +344,25 @@ async function fileToBuiltUrl(
   const file = cleanUrl(id)
   const content = await fsp.readFile(file)
 
-  if (shouldInline == null) {
-    shouldInline =
-      !!config.build.lib ||
-      // Don't inline SVG with fragments, as they are meant to be reused
-      (!(file.endsWith('.svg') && id.includes('#')) &&
-        !file.endsWith('.html') &&
-        content.length < Number(config.build.assetsInlineLimit) &&
-        !isGitLfsPlaceholder(content))
-  }
+  const finalShouldInline = (() => {
+    if (config.build.lib) return true
+    let limit: number
+    if (typeof config.build.assetsInlineLimit === 'function') {
+      const userShouldInline = config.build.assetsInlineLimit(file, content)
+      if (userShouldInline != null) return userShouldInline
+      limit = 4096
+    } else {
+      limit = Number(config.build.assetsInlineLimit)
+    }
+    if (shouldInline != null) return shouldInline
```

Comment:
Should this be before the function call? If we implement `?inline` or `?no-inline`, which gets passed to `shouldInline`, I assume it should always respect the query.

I guess this was added here to appease https://github.com/vitejs/vite/pull/14958#issuecomment-1852751865, but I'm not really sure it's worth it. favicon should always not get inlined as a best practice.

If we do want the function to have the higher priority, I think we should then also move the lib check below, but I'm not sure if this is an ideal default.

Reply:
Yeah I agree that `?inline` or `?no-inline`, if implemented, should have higher prio. I wanted to go in that direction but this API felt simpler to implement and more powerful. 
I did this because it didn't cost much for now, but I agree it complexify things if we later add other ways to inline things, I will change the order